### ISSUE TITLE
Buffers are moved to another device by copy and not in-place

### ISSUE THREAD

**Issue by asanakoy** [pytorch/pytorch#43815]
## 🐛 Bug

Hi guys. The issue was noticed in `pytorch 1.5.1` (it is likely also present in `1.6.0`). 
I'm not sure is it's a bug or a feature.
I have  my own nn.Module where I register some parameters and buffers (call `register_buffer()` on some torch.Tensor-s). 
When I move the module to another device using `m.to(device)`  or `m.cuda()` then the Parameters are moved in-place (so all references to parameters remain valid). However, buffers are copied  to the new device (but nevertheless, module attributes pointing to buffers are updated). This lead to confusion in case one stores references to the buffers in a dictionary.

Example code:
```
import torch
import torch.nn as nn

class Foo(nn.Module):
  def __init__(self):
    super().__init__()
    self.data = {
      'a': torch.Tensor([1]),
      'b': torch.Tensor([1,2,3]),
    }
    for name in self.data:
      self.register_buffer('data_' + name, self.data[name])

m = Foo()
print("m.data['a'] is m.data_a:", m.data['a'] is m.data_a)
print(m.data['a'].device, m.data_a.device)
m.cuda()
print("m.data['a'] is m.data_a:", m.data['a'] is m.data_a)
print(m.data['a'].device, m.data_a.device)

```

It will print:
```
m.data['a'] is m.data_a: True
cpu cpu
m.data['a'] is m.data_a: False
cpu cuda:0
```

When I call `my_module.cuda()` or `my_module.to('gpu:0')` all the registered parameters are moved to the gpu, however buffers in the dictionary are still on CPU. 
So what basically happens is that the objects in the buffers are copied when moving to a new device and `self.data` still contains old references.
**In case of parameters, this wouldn't happen**. And I find in counter-intuitive. 

--------------

Similar Example for parameters:
Example code:
```
import torch
import torch.nn as nn

class FooWithParams(nn.Module):
  def __init__(self):
    super().__init__()
    self.data = {
      'a': nn.Parameter(torch.Tensor([1])),
      'b': nn.Parameter(torch.Tensor([1,2,3])),
    }
    for name in self.data:
      self.register_parameter('data_' + name, self.data[name])

m = FooWithParams()
print("m.data['a'] is m.data_a:", m.data['a'] is m.data_a)
print(m.data['a'].device, m.data_a.device)
m.cuda()
print("m.data['a'] is m.data_a:", m.data['a'] is m.data_a)
print(m.data['a'].device, m.data_a.device)

```

It works as expected:
```
m.data['a'] is m.data_a: True
cpu cpu
m.data['a'] is m.data_a: True
cuda:0 cuda:0
```

So I guess, it would be nice to make he behaviour of buffers similar to the Parameters. Or at least this should be depicted in the docs at the very least to prevent people to fall into this trap.

**Comment by mariosasko** [comment]
From looking at the [source](https://github.com/pytorch/pytorch/blob/7d517cf96f6d53bbe472cf1404e12b9b75230bb6/torch/nn/modules/module.py#L357), it seems like the model parameters are not getting copied but changed in-place because it would introduce backward-compatibility issues. The current behavior should change in the future so both parameters and buffers will get copied (at the moment, it's controlled by a single flag).

**Comment by asanakoy** [comment]
Yes, I see, one can set `torch.__future__.get_overwrite_module_params_on_conversion(True)`. 
As far as I understood this behaviour is needed to enable moving to XLA devices, right?

**Comment by mariosasko** [comment]
> As far as I understood this behaviour is needed to enable moving to XLA devices, right?

Yes, this [PR](https://github.com/pytorch/pytorch/pull/21613) explains it.

**Comment by ailzhang** [comment]
Hmmm I'll take another look at the original PR and see whether it's still needed.

**Comment by ailzhang** [comment]
cc: @gchanan If I understand the discussion in #21613 correctly, copy and overwrite is desired way so the xla part should be fine. Do you know why we don't do the same to buffers? Thanks!
